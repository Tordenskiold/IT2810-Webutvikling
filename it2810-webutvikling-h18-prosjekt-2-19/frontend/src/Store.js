import { decorate, observable, action } from 'mobx';
import axios from 'axios';

class Store {
  searchParams = {
    query: null,
    after: null,
    before: null,
    author: null,
    sortby: null,
    page: 0
  };

  authors = [];

  results = [];

  displayNext = true;

  inspectVisible = false;

  inspectElement = {};

  // The function saves data in the search parameters
  onSelect = (stateName, page, data) => {
    this.searchParams.page = page;
    this.searchParams[stateName] = data;
    this.databaseSearch();
  };

  // The function changes of the pages
  changePage(diff) {
    if (this.searchParams.page + diff >= 0) {
      this.searchParams.page = this.searchParams.page + diff;
      this.databaseSearch();
    }
  }

  // The function adds the data into results array
  pushData(data) {
    data.forEach((entry) => {
      this.results.push(entry);
    });
  }

  // The function replaces the data into results array
  setData(data) {
    this.results = data;
  }

  // The function asks the api and takes the results from the querry and saves the data into
  // results array
  databaseSearch = () => {
    // http://localhost:3000/api/entries/search?query=test&after=2000-09-21&before=2060-09-21&author=rosvik&sortby=author&page=1

    const {
      query, after, before, author, sortby, page
    } = this.searchParams;
    const urlWithParams = new URL('http://localhost:3000/api/entries/search');
    if (query) urlWithParams.searchParams.append('query', query);
    if (after) urlWithParams.searchParams.append('after', after);
    if (before) urlWithParams.searchParams.append('before', before);
    if (author) urlWithParams.searchParams.append('author', author);
    if (sortby) urlWithParams.searchParams.append('sortby', sortby);
    if (page) urlWithParams.searchParams.append('page', page);
    axios.get(urlWithParams.href).then((response) => {
      // this.setData(response.data);
      if (page === 0) {
        this.setData(response.data);
      } else {
        this.pushData(response.data);
      }
      // Show or hide the 'more results' button
      this.displayNext = response.data.length === 5;
    });
  };

  // Gets the authors from the API
  getAuthors = () => {
    const url = 'http://it2810-19.idi.ntnu.no:3000/api/authors';
    axios.get(url).then((response) => {
      const newAuthors = response.data.map(val => ({ value: val.author, label: val.author }));
      newAuthors.unshift({ value: '', label: 'All authors' });
      this.authors = newAuthors;
    });
  };
}

// Defines which variables will be available for different components
decorate(Store, {
  searchParams: observable,
  val: observable,
  results: observable,
  displayNext: observable,
  inspectVisible: observable,
  authors: observable,
  changePage: action,
  pushData: action,
  setData: action,
  onSelect: action
});

export default Store;
