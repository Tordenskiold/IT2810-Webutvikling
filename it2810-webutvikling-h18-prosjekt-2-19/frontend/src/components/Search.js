/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/forbid-prop-types */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row } from 'simple-flexbox';
import { observer } from 'mobx-react';
import SearchBar from './Searchbar';
import Dropdown from './Dropdown';
import Calendar from './Calendar';
import '../css/App.css';

// Search function for the website
class Search extends Component {
  componentDidMount() {
    const { store } = this.props;
    store.getAuthors();
  }

  // Searches for specific querry, takes in the input form the user and saves it to the store
  onQuerySearch = query => this.props.store.onSelect('query', 0, query);

  // Searches for specific users, takes in the input form the user and saves it to the store
  // Drop Down menu parameter
  onSelectUser = user => this.props.store.onSelect('author', 0, user);

  // Searches for specific sorting parameters, takes in the input form the user and saves it to
  // the store
  // Drop Down menu parameter
  onSelectOrder = order => this.props.store.onSelect('sortby', 0, order);

  // Search parameter for the date
  onSelectBefore = timestamp => this.props.store.onSelect('before', 0, timestamp);

  // Search parameter for the date
  onSelectAfter = timestamp => this.props.store.onSelect('after', 0, timestamp);

  // Search parameter for the page
  onSelectPage = page => this.props.store.onSelect('page', this.props.store.searchParams.page + page, 'none');

  render() {
    const { store } = this.props;
    const { authors } = store;

    let authorOptions = [];
    if (authors) authorOptions = authors;

    // Drop Down menu parameters
    const orderBy = [
      { value: 'timestamp', label: 'Date' },
      { value: 'author', label: 'User' },
      { value: 'name', label: 'Title' }
    ];

    // renders the searchbar
    return (
      <div className="App">
        <SearchBar onSearch={this.onQuerySearch} />
        <Row horizontal="center" className="Row">
          <div className="firstDiv">
            <label>From </label>
            <Calendar onSelect={this.onSelectAfter} />
          </div>
          <div className="myDiv">
            <label>to </label>
            <Calendar onSelect={this.onSelectBefore} />
          </div>
          <label className="myLabel">User</label>
          <div id="dropdown1" className="myDivDropdown">
            <Dropdown options={authorOptions} onSelect={this.onSelectUser} />
          </div>
          <label className="myLabel">Sort by</label>
          <div id="dropdown2" className="myDivDropdown">
            <Dropdown options={orderBy} onSelect={this.onSelectOrder} />
          </div>
        </Row>
      </div>
    );
  }
}

observer(Search);
Search.propTypes = {
  store: PropTypes.any.isRequired
};

export default Search;
