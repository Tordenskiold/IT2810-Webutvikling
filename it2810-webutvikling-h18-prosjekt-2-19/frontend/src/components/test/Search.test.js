import React from 'react';
import Search from '../Search';
import Store from '../../Store';
import renderer from 'react-test-renderer';

//Jest test for Search.js and Store.js
//Creates a snapshot test which tests the rendering of the search using the elements from Store.js
it('renders without crashing', () => {
  const appstore = new Store();
  const component = renderer.create(
    <Search store={appstore} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
