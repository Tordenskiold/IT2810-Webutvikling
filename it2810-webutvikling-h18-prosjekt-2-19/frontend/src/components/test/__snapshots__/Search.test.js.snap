// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="App"
>
  <div
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "backgroundColor": "#ffffff",
        "borderRadius": 2,
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "display": "flex",
        "fontFamily": "Roboto, sans-serif",
        "height": 48,
        "justifyContent": "space-between",
        "margin": "0 auto",
        "marginTop": "40px",
        "maxWidth": 800,
        "muiPrepared": true,
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "auto 16px",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "inline-block",
            "muiPrepared": true,
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <div
          className="MainSearch"
          style={
            Object {
              "backgroundColor": "transparent",
              "cursor": "auto",
              "display": "inline-block",
              "fontFamily": "Roboto, sans-serif",
              "fontSize": 16,
              "height": 48,
              "lineHeight": "24px",
              "muiPrepared": true,
              "position": "relative",
              "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "bottom": 12,
                "color": "rgba(0, 0, 0, 0.3)",
                "muiPrepared": true,
                "opacity": 1,
                "position": "absolute",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              }
            }
          >
            Search
          </div>
          <input
            autoComplete="off"
            disabled={false}
            id="1"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            open={false}
            spellCheck={false}
            style={
              Object {
                "WebkitOpacity": 1,
                "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                "appearance": "textfield",
                "backgroundColor": "rgba(0,0,0,0)",
                "border": "none",
                "color": "rgba(0, 0, 0, 0.87)",
                "cursor": "inherit",
                "font": "inherit",
                "height": "100%",
                "muiPrepared": true,
                "outline": "none",
                "padding": 0,
                "position": "relative",
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </div>
        <div
          style={
            Object {
              "display": "none",
            }
          }
        />
      </div>
    </div>
    <button
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOut={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "background": "none",
          "border": 10,
          "boxSizing": "border-box",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": 0,
          "fontWeight": "inherit",
          "height": 48,
          "margin": 0,
          "marginRight": -48,
          "muiPrepared": true,
          "opacity": 0.54,
          "outline": "none",
          "overflow": "visible",
          "padding": 12,
          "position": "relative",
          "textDecoration": "none",
          "transform": "scale(1, 1)",
          "transition": "transform 200ms cubic-bezier(0.4, 0.0, 0.2, 1)",
          "verticalAlign": null,
          "width": 48,
        }
      }
      tabIndex={0}
      type="button"
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "inline-block",
              "fill": "#9e9e9e",
              "height": 24,
              "muiPrepared": true,
              "opacity": 1,
              "transition": "opacity 200ms cubic-bezier(0.4, 0.0, 0.2, 1)",
              "userSelect": "none",
              "width": 24,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
    </button>
    <button
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOut={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "background": "none",
          "border": 10,
          "boxSizing": "border-box",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": 0,
          "fontWeight": "inherit",
          "height": 48,
          "margin": 0,
          "muiPrepared": true,
          "opacity": 0.54,
          "outline": "none",
          "overflow": "visible",
          "padding": 12,
          "position": "relative",
          "textDecoration": "none",
          "transform": "scale(0, 0)",
          "transition": "transform 200ms cubic-bezier(0.4, 0.0, 0.2, 1)",
          "verticalAlign": null,
          "width": 48,
        }
      }
      tabIndex={0}
      type="button"
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "inline-block",
              "fill": "#9e9e9e",
              "height": 24,
              "muiPrepared": true,
              "opacity": 0,
              "transition": "opacity 200ms cubic-bezier(0.4, 0.0, 0.2, 1)",
              "userSelect": "none",
              "width": 24,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    className="Row"
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "noWrap",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="firstDiv"
    >
      <label>
        From 
      </label>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <button
            className="example"
            onClick={[Function]}
            type="button"
          >
            11/16/2018
          </button>
        </div>
      </div>
    </div>
    <div
      className="myDiv"
    >
      <label>
        to 
      </label>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <button
            className="example"
            onClick={[Function]}
            type="button"
          >
            11/16/2018
          </button>
        </div>
      </div>
    </div>
    <label
      className="myLabel"
    >
      User
    </label>
    <div
      className="myDivDropdown"
      id="dropdown1"
    >
      <div
        className="css-10nd86i"
        onKeyDown={[Function]}
      >
        <div
          className="css-vj8t7z"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-2-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="css-d8oujb"
            />
            <div
              aria-hidden="true"
              className="css-1ep9fjw"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <label
      className="myLabel"
    >
      Sort by
    </label>
    <div
      className="myDivDropdown"
      id="dropdown2"
    >
      <div
        className="css-10nd86i"
        onKeyDown={[Function]}
      >
        <div
          className="css-vj8t7z"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-3-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="css-d8oujb"
            />
            <div
              aria-hidden="true"
              className="css-1ep9fjw"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-19bqh2r"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
